wasm/jit/jit_amd64.go

```diff
@@ -4594,66 +4569,695 @@ func (c *amd64Compiler) allocateRegister(t generalPurposeRegisterType) (reg int1
+	readCurrentCallFrameReturnAddress.From.Index = currentCallFrameAddressRegister
+	readCurrentCallFrameReturnAddress.From.Scale = 1
+	readCurrentCallFrameReturnAddress.From.Offset = -(callFrameDataSize - callFrameReturnAddressOffset)
+	c.addInstruction(readCurrentCallFrameReturnAddress)
+
+	readInstructionAddressCompletionCallback := c.readInstructionAddress(ripRegister)
+
+	// We are ready to store the return address (in ripRegister) to engine.callFrameStack[engine.callStackFramePointer-1].
+	moveRIP := c.newProg()
+	moveRIP.As = x86.AMOVQ
+	moveRIP.To.Type = obj.TYPE_MEM
+	moveRIP.To.Reg = currentCallFrameAddressRegister
+	moveRIP.To.Offset = callFrameReturnAddressOffset
+	moveRIP.From.Type = obj.TYPE_REG
+	moveRIP.From.Reg = ripRegister
+	c.addInstruction(moveRIP)
+
+	// Emit the return instruction.
+	c.exit(jitStatus)
+
+	nopAfterReturnInst := c.newProg()
+	nopAfterReturnInst.As = obj.ANOP
+	c.addInstruction(nopAfterReturnInst)
+
+	readInstructionAddressCompletionCallback(moveRIP, nopAfterReturnInst)
+
+	// They were temporarily used, so we mark them unused.
+	c.locationStack.markRegisterUnused(regs...)
 	return nil
 }
 
+// readInstructionAddress adds the instruction to read the target instruction's absolute address into the destinationRegister.
+// At the callsite, the returned callback must be called with two arguments:
+// instructionAfterReadInstructionAddress ... is the instruction added *right after* calling this readInstructionAddress function.
+// addressAcquisitionTargetInstruction ... is the instruction we want to get the absolute address.
+func (c *amd64Compiler) readInstructionAddress(destinationRegister int16) func(instructionAfterReadInstructionAddress, addressAcquisitionTargetInstruction *obj.Prog) {
+	readInstructionAddressAfterReturn := c.newProg()
+	readInstructionAddressAfterReturn.As = x86.ALEAQ
+	readInstructionAddressAfterReturn.To.Reg = destinationRegister
+	readInstructionAddressAfterReturn.To.Type = obj.TYPE_REG
+	readInstructionAddressAfterReturn.From.Type = obj.TYPE_MEM
+	// We use place holder here as we don't yet know at this point the offset of the first instruction
+	// after return instruction.
+	readInstructionAddressAfterReturn.From.Offset = 0xffff
+	// Since the assembler cannot directly emit "LEA foo [rip + bar]", we use the some hack here:
+	// We intentionally use x86.REG_BP here so that the resulting instruction sequence becomes
+	// exactly the same as "LEA foo [rip + bar]" except the most significant bit of the third byte.
+	// We do the rewrite in onGenerateCallbacks which is invoked after the assembler emitted the code.
+	readInstructionAddressAfterReturn.From.Reg = x86.REG_BP
+	c.addInstruction(readInstructionAddressAfterReturn)
+
+	return func(instructionAfterReadInstructionAddress, addressAcquisitionTargetInstruction *obj.Prog) {
+		c.onGenerateCallbacks = append(c.onGenerateCallbacks, func(code []byte) error {
+			// See the comment at readInstructionAddressAfterReturn.From.Offset.
```

Comment:
extract this to a function so that you can use godoc links (note only direct symbols link, not nested)

Reply:
I don't think we could extract this as a function because  this is the closure on readInstructionAddressAfterReturn..

Reply:
at least simplified! https://github.com/tetratelabs/wazero/pull/161/commits/6bff87ebff3d103eb6d68cb384a4d868703f123d